<?xml version="1.0"?>
<launch>
  <!-- Load hardware configuration -->
  <rosparam command="load" file="$(find orca_hw_interface)/config/hardware.yaml"/>
  
  <!-- Start pressure sensor node -->
  <node name="pressure_sensor_node" pkg="orca_hw_interface" type="pressure_sensor_node" output="screen">
    <param name="i2c_bus" value="$(arg i2c_bus)" if="$(eval arg('i2c_bus') != '')"/>
    <param name="i2c_address" value="$(arg i2c_address)" if="$(eval arg('i2c_address') != 0)"/>
    <param name="fluid_density" value="$(arg fluid_density)" if="$(eval arg('fluid_density') != 0.0)"/>
    <param name="update_rate" value="$(arg update_rate)" if="$(eval arg('update_rate') != 0.0)"/>
  </node>
  
  <!-- Start camera configuration node -->
  <node name="camera_config" pkg="orca_hw_interface" type="camera_config.py" output="screen">
    <rosparam param="camera_names" subst_value="True">$(arg camera_names)</rosparam>
    <rosparam param="camera_urls" subst_value="True">$(arg camera_urls)</rosparam>
    <rosparam param="camera_frames" subst_value="True">$(arg camera_frames)</rosparam>
  </node>
  
  <!-- Start MAVROS -->
  <include file="$(find mavros)/launch/node.launch">
    <arg name="pluginlists_yaml" value="$(find mavros)/launch/apm_pluginlists.yaml"/>
    <arg name="config_yaml" value="$(find mavros)/launch/apm_config.yaml"/>
    <arg name="fcu_url" value="$(arg fcu_url)"/>
    <arg name="gcs_url" value="$(arg gcs_url)"/>
    <arg name="tgt_system" value="$(arg tgt_system)"/>
    <arg name="tgt_component" value="$(arg tgt_component)"/>
  </include>
  
  <!-- Start RealSense camera -->
  <include file="$(find realsense2_camera)/launch/rs_camera.launch">
    <arg name="enable_pointcloud" value="true"/>
    <arg name="align_depth" value="true"/>
    <arg name="filters" value="pointcloud"/>
  </include>
  
  <!-- Start front camera -->
  <node name="front_camera" pkg="usb_cam" type="usb_cam_node" output="screen">
    <param name="video_device" value="/dev/video0"/>
    <param name="image_width" value="1920"/>
    <param name="image_height" value="1080"/>
    <param name="pixel_format" value="yuyv"/>
    <param name="camera_frame_id" value="front_camera_link"/>
    <param name="camera_name" value="front_camera"/>
    <param name="io_method" value="mmap"/>
    <param name="framerate" value="30"/>
  </node>
  
  <!-- Start rear camera -->
  <node name="rear_camera" pkg="usb_cam" type="usb_cam_node" output="screen">
    <param name="video_device" value="/dev/video1"/>
    <param name="image_width" value="640"/>
    <param name="image_height" value="480"/>
    <param name="pixel_format" value="yuyv"/>
    <param name="camera_frame_id" value="rear_camera_link"/>
    <param name="camera_name" value="rear_camera"/>
    <param name="io_method" value="mmap"/>
    <param name="framerate" value="30"/>
  </node>
</launch>
